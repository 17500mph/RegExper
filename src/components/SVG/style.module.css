@import url('../../globals.module.css');

.render {
  width: 100%;
  background: var(--color-white);
  box-sizing: border-box;
  overflow: auto;
  margin: var(--spacing-margin) 0;

  & svg {
    display: block;
    transform: scaleZ(1); /* Move to separate render layer in Chrome */
    margin: 0 auto;
  }
}
