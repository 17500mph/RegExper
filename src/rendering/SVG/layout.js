import layout from 'layout';

const layoutSVG = data => {
  const child = layout(data.children[0]);

  data.props.innerWidth = child.box.width;
  data.props.innerHeight = child.box.height;

  return data;
};

export default layoutSVG;
