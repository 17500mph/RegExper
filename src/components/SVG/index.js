import React from 'react';

import PlaceholderIcon from 'react-feather/dist/icons/file-text';

import style from './style.module.css';

import AppContext from 'components/AppContext';

class SVG extends React.PureComponent {
  static contextType = AppContext

  svgContainer = React.createRef()

  provideSVGData() {
    if (!this.svgContainer.current) {
      return;
    }

    const svg = this.svgContainer.current.querySelector('svg');
    this.context.setSvgMarkup({
      svg: svg.outerHTML,
      width: svg.getAttribute('width'),
      height: svg.getAttribute('height')
    });
  }

  componentDidMount() {
    this.provideSVGData();
  }

  componentDidUpdate() {
    this.provideSVGData();
  }

  render() {
    // Demo rendering for now
    return <div className={ style.render } ref={ this.svgContainer }>
      <PlaceholderIcon />
    </div>;
  }
}

export default SVG;
