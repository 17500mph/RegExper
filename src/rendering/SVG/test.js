import React from 'react';
import { shallow } from 'enzyme';

import SVG, { layout } from 'rendering/SVG';

describe('SVG', () => {
  test('rendering', () => {
    const component = shallow(
      <SVG width={ 100 } height={ 20 }>Content</SVG>
    );
    expect(component).toMatchSnapshot();
  });

  test('layout', () => {
    const processed = layout({
      type: 'SVG',
      children: [
        {
          type: 'Text',
          box: {
            width: 100,
            height: 20
          }
        }
      ]
    });
    expect(processed).toMatchSnapshot();
  });
});
