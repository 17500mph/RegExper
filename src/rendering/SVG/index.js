import React from 'react';
import PropTypes from 'prop-types';

import * as style from './style';

const namespaceProps = {
  'xmlns': 'http://www.w3.org/2000/svg',
  'xmlns:cc': 'http://creativecommons.org/ns#',
  'xmlns:rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#'
};
/* eslint-disable max-len */
const metadata = `<rdf:rdf>
  <cc:license rdf:about="http://creativecommons.org/licenses/by/3.0/">
    <cc:permits rdf:resource="http://creativecommons.org/ns#Reproduction"></cc:permits>
    <cc:permits rdf:resource="http://creativecommons.org/ns#Distribution"></cc:permits>
    <cc:requires rdf:resource="http://creativecommons.org/ns#Notice"></cc:requires>
    <cc:requires rdf:resource="http://creativecommons.org/ns#Attribution"></cc:requires>
    <cc:permits rdf:resource="http://creativecommons.org/ns#DerivativeWorks"></cc:permits>
  </cc:license>
</rdf:rdf>`;
/* eslint-enable max-len */

class SVG extends React.PureComponent {
  static propTypes = {
    onReflow: PropTypes.func,
    children: PropTypes.node,
    padding: PropTypes.number,
    innerWidth: PropTypes.number,
    innerHeight: PropTypes.number
  }

  static defaultProps = {
    padding: 10
  }

  state = {
    width: 0,
    height: 0
  }

  render() {
    const { padding, innerWidth, innerHeight, children } = this.props;

    const width = Math.round(innerWidth + 2 * padding);
    const height = Math.round(innerHeight + 2 * padding);

    const svgProps = {
      width,
      height,
      viewBox: [0, 0, width, height].join(' '),
      style: style.image,
      ...namespaceProps
    };

    return <svg { ...svgProps }>
      <metadata dangerouslySetInnerHTML={{ __html: metadata }}></metadata>
      <g transform={ `translate(${ padding } ${ padding })` }>
        { children }
      </g>
    </svg>;
  }
}

export default SVG;
