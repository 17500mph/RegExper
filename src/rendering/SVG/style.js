import { white } from 'rendering/style';

export const image = {
  backgroundColor: white
};
