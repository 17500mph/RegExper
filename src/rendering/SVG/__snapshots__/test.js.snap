// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVG layout 1`] = `
Object {
  "box": Object {
    "height": 40,
    "width": 120,
  },
  "children": Array [
    Object {
      "box": Object {
        "height": 20,
        "width": 100,
      },
      "type": "Text",
    },
  ],
  "props": Object {
    "height": 40,
    "width": 120,
  },
  "type": "SVG",
}
`;

exports[`SVG rendering 1`] = `
<svg
  height={20}
  style={
    Object {
      "backgroundColor": "#fff",
    }
  }
  viewBox="0 0 100 20"
  width={100}
  xmlns="http://www.w3.org/2000/svg"
  xmlns:cc="http://creativecommons.org/ns#"
  xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
>
  <metadata
    dangerouslySetInnerHTML={
      Object {
        "__html": "<rdf:rdf>
  <cc:license rdf:about=\\"http://creativecommons.org/licenses/by/3.0/\\">
    <cc:permits rdf:resource=\\"http://creativecommons.org/ns#Reproduction\\"></cc:permits>
    <cc:permits rdf:resource=\\"http://creativecommons.org/ns#Distribution\\"></cc:permits>
    <cc:requires rdf:resource=\\"http://creativecommons.org/ns#Notice\\"></cc:requires>
    <cc:requires rdf:resource=\\"http://creativecommons.org/ns#Attribution\\"></cc:requires>
    <cc:permits rdf:resource=\\"http://creativecommons.org/ns#DerivativeWorks\\"></cc:permits>
  </cc:license>
</rdf:rdf>",
      }
    }
  />
  <g
    transform="translate(10 10)"
  >
    Content
  </g>
</svg>
`;
